### ISSUE TITLE
Bug: CLI option `--plugin` cannot load CommonJS plugin in flat config mode

### ISSUE THREAD

**Issue by fasttime** [eslint/eslint#18159]
### Environment

Node version: v21.6.2
npm version: v10.2.4
Local ESLint version: v9.0.0-beta.1 (Currently used)
Global ESLint version: Not found
Operating System: darwin 23.3.0

### What parser are you using?

Default (Espree)

### What did you do?

In the eslint repo directory, I used the CLI to run a single rule from a plugin on a file:

```shell
npx eslint --no-config-lookup --plugin n --rule 'n/no-sync: error' --parser-options 'sourceType: commonjs' Makefile.js
```

### What did you expect to happen?

The command should report problems in the specified file.
For comparison, this is the output I see in a zsh terminal when using eslintrc config:

```console
$ ESLINT_USE_FLAT_CONFIG=false npx eslint --no-eslintrc --plugin n --rule 'n/no-sync: error' --parser-options 'ecmaVersion: latest, sourceType: commonjs' Makefile.js

.../Makefile.js
   77:49  error  Unexpected sync method: 'readFileSync'   n/no-sync
  103:22  error  Unexpected sync method: 'readFileSync'   n/no-sync
  296:5   error  Unexpected sync method: 'writeFileSync'  n/no-sync
  401:30  error  Unexpected sync method: 'readFileSync'   n/no-sync
  424:28  error  Unexpected sync method: 'readdirSync'    n/no-sync
  821:13  error  Unexpected sync method: 'execFileSync'   n/no-sync
  821:13  error  Unexpected sync method: 'execFileSync'   n/no-sync
  825:9   error  Unexpected sync method: 'execFileSync'   n/no-sync

✖ 8 problems (8 errors, 0 warnings)

(node:10816) ESLintRCWarning: You are using an eslintrc configuration file, which is deprecated and support will be removed in v10.0.0. Please migrate to an eslint.config.js file. See https://eslint.org/docs/latest/use/configure/migration-guide for details.
(Use `node --trace-warnings ...` to show where the warning was created)
```

### What actually happened?

The command fails with an error for not finding the specified rule:

```console
$ npx eslint --no-config-lookup --plugin n --rule 'n/no-sync: error' --parser-options 'sourceType: commonjs' Makefile.js

(node:3528) ESLintIgnoreWarning: The ".eslintignore" file is no longer supported. Switch to using the "ignores" property in "eslint.config.js": https://eslint.org/docs/latest/use/configure/migration-guide#ignoring-files
(Use `node --trace-warnings ...` to show where the warning was created)

Oops! Something went wrong! :(

ESLint: 9.0.0-beta.1

TypeError: Key "rules": Key "n/no-sync": Could not find "no-sync" in plugin "n". Did you mean "@/no-sync"?
    at throwRuleNotFoundError (.../eslint/lib/config/rule-validator.js:66:11)
    at RuleValidator.validate (.../eslint/lib/config/rule-validator.js:147:17)
    at [finalizeConfig] (.../eslint/lib/config/flat-config-array.js:211:23)
    at FlatConfigArray.getConfig (.../eslint/node_modules/@humanwhocodes/config-array/api.js:938:55)
    at FlatConfigArray.isFileIgnored (.../eslint/node_modules/@humanwhocodes/config-array/api.js:962:15)
    at .../eslint/lib/eslint/eslint-helpers.js:514:38
    at Array.forEach (<anonymous>)
    at findFiles (.../eslint/lib/eslint/eslint-helpers.js:503:11)
    at async ESLint.lintFiles (.../eslint/lib/eslint/eslint.js:834:27)
    at async Object.execute (.../eslint/lib/cli.js:442:23)
```

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/stackblitz-starters-h9q29a?file=index.js

### Participation

- [X] I am willing to submit a pull request for this issue.

### Additional comments

This issue only occurs when a CommonJS plugin is loaded from the CLI in flat config. In my understanding, the problem is that when a CommonJS module is loaded with the dynamic `import` syntax, it gets transpiled into an ESM module which is only guaranteed to have a `default` export.

https://github.com/eslint/eslint/blob/bde51055530d4a71bd9f48c90ed7de9c0b767d86/lib/cli.js#L162

So the rules exported by the CommonJS plugin will be accessible as `default.rules` from the resolved module object. This is in contrast with the way ESM modules are loaded, where all exports are made available as top level properties.

**Comment by mdjermanovic** [comment]
I can reproduce this. Also, it looks like we don't have any CLI tests for the `--plugin` option in flat config mode.

**Comment by mdjermanovic** [comment]
I think there will also be a similar problem with ESM plugins that `export default { rules, ... }`, which probably makes more sense than named exports given that a plugin is considered to be an object, so perhaps we should always do `plugins[shortName] = (await importer.import(longName)).default`, and also require that ESM plugins must use `export default`?

**Comment by Rec0iL99** [comment]
ping @eslint/eslint-team

**Comment by nzakas** [comment]
Thanks @Rec0iL99 for the ping!

@mdjermanovic please remember to ping @eslint/eslint-team when moving an issue to "Feedback Needed".

> so perhaps we should always do `plugins[shortName] = (await importer.import(longName)).default`

That makes sense to me. Let's fix it and add tests. 👍 

> require that ESM plugins must use `export default`

I don't think we have a way to enforce this given that it's only the CLI `--plugin` option that really cares. (I'm honestly not sure how useful `--plugin` is in the real world, so it's not heartbreaking to be a bit more strict about its use.) We do [document](https://eslint.org/docs/next/extend/plugins) using default exports in the docs, though, so probably not a big deal.
